src/repl/scala/tools/nsc/interpreter/LoopCommands.scala

```diff
@@ -116,22 +117,34 @@ trait LoopCommands { self: { def echo(msg: String): Unit } =>
   }
 
   import Completion.Candidates
-
-  def colonCompletion(line: String, cursor: Int): Completion = line.trim match {
-    case commandish(name @ CommandMatch(cmd), rest) =>
-      if (name.length > cmd.name.length) cmd.completion
-      else
-        new Completion {
-          def resetVerbosity(): Unit = ()
-          def complete(buffer: String, cursor: Int) = Candidates(cursor - name.length + 1, List(cmd.name))
+  import scala.util.matching.Regex.Groups
+
+  def colonCompletion(line: String, cursor: Int): Completion =
+    commandish.findFirstMatchIn(line) match {
+      case Some(m) =>
+        val Groups(name0, rest) = m
+        val name = name0 take cursor
+        val cmds = matchingCommands(name)
+        val cursorAtName = cursor <= name.length
+        cmds match {
+          case Nil                            => NoCompletion
+          case cmd :: Nil if !cursorAtName    => cmd.completion
+          case cmd :: Nil if cmd.name == name => NoCompletion
+          case cmd :: Nil =>
+            val suffix = cmd.name.substring(cursor - 1)   // account for colon
+            val completion = if (cmd.isInstanceOf[NullaryCmd] || cursor < line.length) suffix else suffix + " "
+            new Completion {
+              def resetVerbosity(): Unit = ()
+              def complete(buffer: String, cursor: Int) = Candidates(cursor, List(completion))
+            }
+          case cmd :: rest =>
+            new Completion {
+              def resetVerbosity(): Unit = ()
+              def complete(buffer: String, cursor: Int) = Candidates(cursor, cmds.map(_.name))
```

Comment:
Should this be `Candidates(cursor, cmds.map(_.name.drop(cursor - 1)))`?

As it is, `:imp` completes to `:impimp` with suggestions for `imports` and `implicits`.

Reply:
I think I had that version at some point. Lemme check.